Sources/LLVM/Function+Attributes.swift

```diff
@@ -0,0 +1,262 @@
+  /// return value).
+  case signext
+  /// This indicates that this parameter or return value should be treated in a
+  /// special target-dependent fashion while emitting code for a function call
+  /// or return.
+  case inreg
+  /// This indicates that the pointer parameter should really be passed by
+  /// value to the function.
+  case byval
+  /// The inalloca argument attribute allows the caller to take the address of
+  /// outgoing stack arguments
+  case inalloca
+  /// This indicates that the pointer parameter specifies the address of a
+  /// structure that is the return value of the function in the source program.
+  case sret
+  /// This indicates that the pointer value may be assumed by the optimizer to
+  /// have the specified alignment.
+  case align
+  /// This indicates that objects accessed via pointer values based on the
+  /// argument or return value are not also accessed, during the execution of
+  /// the function, via pointer values not based on the argument or return
+  /// value.
+  case noalias
+  /// This indicates that the callee does not make any copies of the pointer
+  /// that outlive the callee itself.
+  case nocapture
+  /// This indicates that the pointer parameter can be excised using the
+  /// trampoline intrinsics.
+  case nest
+  /// This indicates that the function always returns the argument as its
+  /// return value.
+  case returned
+  /// This indicates that the parameter or return pointer is not null.
+  case nonnull
+  /// This indicates that the parameter or return pointer is dereferenceable.
+  case dereferenceable
+  /// This indicates that the parameter or return value isn’t both non-null and
+  /// non-dereferenceable (up to `n` bytes) at the same time.
+  case dereferenceable_or_null
+  /// This indicates that the parameter is the self/context parameter.
+  case swiftself
+  /// This attribute is motivated to model and optimize Swift error handling.
+  case swifterror
+
+  /// ID of the attribute.
+  internal var kindID: UInt32 {
+    return LLVMGetEnumAttributeKindForName(rawValue, rawValue.count)
+  }
+}
+
+extension Function {
+  /// Adds an attribute to the function.
+  ///
+  /// - parameter attr: The attribute to add.
+  /// - parameter value: The optional value of the attribute.
+  public func addAttribute(_ attr: FunctionAttribute, value: UInt64 = 0) {
+    let ctx = LLVMGetModuleContext(LLVMGetGlobalParent(llvm))
+    let attrRef = LLVMCreateEnumAttribute(ctx, attr.kindID, value)
+    LLVMAddAttributeAtIndex(llvm, 0, attrRef)
+  }
+
+  /// Adds an attribute to the given function's parameter.
+  ///
+  /// - parameter attr: The attribute to add.
+  /// - parameter value: The optional value of the attribute.
+  /// - parameterIndex: The index of the parameter to which add the
+  ///   attribute, starting from 0.
+  public func addParameterAttribute(_ attr: ParameterAttribute, value: UInt64 = 0,
+                                    for parameterIndex: LLVMAttributeIndex) {
```

Comment:
We have a policy of never exposing LLVM's types to the user.  I have an idea of how we can lift this into an enum.

Reply:
Please use `Int` here.